File: mcmc.py
import math
import numpy
import random

from recorder import Recorder
from tree import Tree


def kingman_mcmc(tree, recorder):
    acceptance_prob_spr = 0
    acceptance_prob_times = 0
    log_likelihood = tree.log_likelihood()
    steps = int(recorder.tables.sequence_length) - 1

    for i in range(steps):
        child = tree.sample_leaf()
        parent = tree.parent[child]
        sib = tree.sibling(child)
        new_sib = sib
        if tree.sample_size > 2:
            while new_sib in [child, parent, sib]:
                new_sib = tree.sample_node()
        new_time = tree.sample_reattach_time(child, new_sib)
        alpha = -tree.log_reattach_density(child, new_sib, new_time)
        old_time = tree.time[parent]
        tree.detach_reattach(child, new_sib, new_time)
        alpha += tree.log_reattach_density(child, sib, old_time)
        proposal_log_likelihood = tree.log_likelihood()
        alpha += proposal_log_likelihood - log_likelihood
        if math.log(random.random()) < alpha:
            log_likelihood = proposal_log_likelihood
            acceptance_prob_spr += 1 / steps
        else:
            tree.detach_reattach(child, sib, old_time)

        old_times = tree.time.copy()
        alpha = -tree.resample_times()
        proposal_log_likelihood = tree.log_likelihood()
        alpha += tree.log_resample_times_density(old_times)
        alpha += proposal_log_likelihood - log_likelihood
        if math.log(random.random()) < alpha:
            log_likelihood = proposal_log_likelihood
            acceptance_prob_times += 1 / steps
        else:
            tree.time = old_times.copy()

        recorder.append_tree(tree)

    return [acceptance_prob_spr, acceptance_prob_times]

File: C/mcmc.c
#include "mcmc.h"
#include "recorder.h"
#include "tree.h"
#include <tskit.h>

void kingman_mcmc(recorder *r, tree *t, gsl_rng *gen) {
  double llik = log_likelihood(t);
  int child, parent, sib, new_sib;
  double new_time, old_time, alpha;
  double proposal_llik;
  double *old_times = malloc((2 * t->sample_size - 1) * sizeof(double));
  int steps = r->tables.sequence_length;
  for (int i = 1; i < steps; i++) {
    child = sample_leaf(t, gen);
    parent = t->parent[child];
    sib = sibling(t, child);
    new_sib = sib;
    if (t->sample_size > 2) {
      while (new_sib == sib || new_sib == parent || new_sib == child) {
        new_sib = sample_node(t, gen);
      }
    }
    new_time = sample_reattach_time(t, child, new_sib, gen);
    alpha = -log_reattach_density(t, child, new_sib, new_time);
    old_time = t->time[parent];
    detach_reattach(t, child, new_sib, new_time);
    alpha += log_reattach_density(t, child, sib, old_time);
    proposal_llik = log_likelihood(t);
    alpha += proposal_llik - llik;
    if (log(gsl_rng_uniform(gen)) < alpha) {
      llik = proposal_llik;
    } else {
      detach_reattach(t, child, sib, old_time);
    }
    for (int j = t->sample_size; j < 2 * t->sample_size - 1; j++) {
      old_times[j] = t->time[j];
    }
    alpha = -resample_times(t, gen);
    proposal_llik = log_likelihood(t);
    alpha += log_resample_times_density(t, old_times);
    alpha += proposal_llik - llik;
    if (log(gsl_rng_uniform(gen)) < alpha) {
      llik = proposal_llik;
    } else {
      for (int j = t->sample_size; j < 2 * t->sample_size - 1; j++) {
        t->time[j] = old_times[j];
      }
    }
    append_tree(r, t);
  }
  free(old_times);
  return;
}
